crates/runtime/src/libcalls.rs

```diff
@@ -409,3 +410,21 @@ pub unsafe extern "C" fn wasmtime_data_drop(vmctx: *mut VMContext, data_index: u
     let instance = (&mut *vmctx).instance();
     instance.data_drop(data_index)
 }
+
+/// Drop a `VMExternRef` in place.
+pub unsafe extern "C" fn wasmtime_drop_externref_in_place(pointer_to_externref: *mut VMExternRef) {
+    ptr::drop_in_place(pointer_to_externref);
```

Comment:
This seems like it may not be right, we don't have a `*mut VMExternRef` here I thought but rather `*mut VMExternData`? Should this be `VMExternData::drop_and_dealloc`?

Reply:
(also does this mean that a test should be added for this too?

Reply:
This isn't a `*mut VMExternData` (which is indeed what a `VMExternRef` is represented with) but actually a pointer to a `VMExternRef`. The `table_addr` instruction evaluates to a pointer to the table element, not the table element itself, and it is the result of the `table_addr` that we pass as an argument to the call to this function.

Reply:
> (also does this mean that a test should be added for this too?

I am 99% sure the spec tests exercise this path, but I can double check, and it might make sense to explicitly ensure that we test this, since it requires `ref_count == 1` on the current table element that will be overridden.

Reply:
Ah sorry I assuming the pseudocode a bit too literally, if that's the case then this is all good!

(although I think this still runs afoul of the issue where the dtor tries to touch the table, so we may not be able to implement it as this for long)

Reply:
> (although I think this still runs afoul of the issue where the dtor tries to touch the table, so we may not be able to implement it as this for long)

Hm this is a good point!

Normally, we would be safe because the table elements are in a `RefCell`, but the JIT code doesn't set the `RefCell`'s borrow flags :-/

The only way I see to avoid this would be to either

* Make a custom version of `RefCell` where the JIT knows how to poke at the borrow flags (not great to add this extra overhead if we can avoid it, plus duplicating a bunch of `std` code)

* Do some kind of deferred destruction, where we queue up ready-to-die references and drain them at known safe points (opens the door for even more confusing/unexpected drop times).

Of the two, I think deferred destruction is probably the more promising approach.

Reply:
We talked a bit on Zulip about how I think by swapping pointers we can sidestep the issue here, but this also makes me think that we should stop using `RefCell` in `Table` internals, instead switching to `UnsafeCell` because the JIT doesn't know about it.